src/lib/reflection/reflect.ts

```diff
@@ -27,33 +30,73 @@ type ReflectionResult = ReflectionResultPlugins | ReflectionResultArtifactGenera
 /**
  * Run the reflection step of Nexus. Get the used plugins and generate the artifacts optionally.
  */
-export function reflect(layout: Layout.Layout, opts: { usedPlugins: true }): Promise<ReflectionResultPlugins>
+export function reflect(
+  layout: Layout.Layout,
+  opts: { usedPlugins: true; onMainThread: true }
+): Promise<ReflectionResultPlugins>
 export function reflect(
   layout: Layout.Layout,
   opts: { artifacts: true }
 ): Promise<ReflectionResultArtifactGeneration>
-export function reflect(
+export async function reflect(
   layout: Layout.Layout,
-  opts: { usedPlugins?: true; artifacts?: true }
+  opts: { usedPlugins?: true; artifacts?: true; onMainThread?: boolean }
 ): Promise<ReflectionResultPlugins | ReflectionResultArtifactGeneration> {
   log.trace('reflection started')
+  if (opts.artifacts) {
+    return runTypegenReflectionAsSubProcess(layout)
+  }
+
+  if (opts.usedPlugins === true && opts.onMainThread === false) {
+    throw new Error('Not implemented. Reflection on plugins needs to be done on the main process for now.')
+  }
+
+  return runPluginsReflectionOnMainThread(layout)
+}
+
+/**
+ * Hack: Plugins should ideally be discovered in a sub-process.
+ * This is temporary until https://github.com/graphql-nexus/nexus/issues/817 is fixed
```

Comment:
Not sure I understand the comment. Isn't this PR fixing that issue?

Reply:
Yeah sorry, I meant #818. Fixed